### PR TITLE
TypeScript 3.7 (take 2)

### PR DESCRIPTION
**Pull request by BPScott**
### WHY are these changes introduced?

#2237 bumped us up to typescript 3.7, but it got reverted in #2501 because it didn't play nice in consuming projects.

This reverts the revert and fixes the issues that caused the problems in consuming projects.

### WHAT is this pull request doing?

Updates to TS 3.7, fixing any type checking issues.

If you have an uninitialized property that then gets referenced (like by
having a getter for it) then TS emits declarations that break in
consuming projects. Convert a bunch of cases where we created getters for non-existant
properties to use boring old method functions, thus allowing consumer builds to work again..

### How to 🎩

Tests / linting pass

Test this in a consuming project:

- Checkout the latest version of polaris-styleguide
- In polaris-react: `yarn run build-consumer polaris-styleguide`
- Then in the styleguide `yarn run build` and see that the build passes

### CODE REVIEW

File: src/components/Filters/Filters.tsx [hunk 1]
```diff
@@ -107,14 +107,6 @@ class Filters extends React.Component<ComposedProps, State> {
   private moreFiltersButtonContainer = createRef<HTMLDivElement>();
   private focusNode = createRef<HTMLDivElement>();
 
-  private get hasAppliedFilters(): boolean {
```
Comment (BPScott) [inline src/components/Filters/Filters.tsx]: ESLint wants us to put private methods below public ones like render().

File: src/components/ResourceList/ResourceList.tsx [hunk 2]
```diff
@@ -329,13 +329,15 @@ class ResourceList extends React.Component<CombinedProps, State> {
     };
   }
 
+  // eslint-disable-next-line @typescript-eslint/member-ordering
```
Comment (BPScott) [inline src/components/ResourceList/ResourceList.tsx]: ESLint wants us to put private methods below public ones like render(). However in this file that'll make the diff incredibly noisy so I've surpassed the warning

File: src/components/ResourceList/components/BulkActions/BulkActions.tsx [hunk 3]
```diff
@@ -169,6 +169,7 @@ class BulkActions extends React.PureComponent<CombinedProps, State> {
     }
   }
 
+  // eslint-disable-next-line @typescript-eslint/member-ordering
```
Comment (BPScott) [inline src/components/ResourceList/components/BulkActions/BulkActions.tsx]: ESLint wants us to put private methods below public ones like render(). However in this file that'll make the diff incredibly noisy so I've surpassed the warning

File: src/components/ResourceList/components/BulkActions/BulkActions.tsx [hunk 4]
```diff
@@ -204,6 +206,8 @@ class BulkActions extends React.PureComponent<CombinedProps, State> {
       polaris: {intl},
     } = this.props;
 
+    const actionSections = this.actionSections();
```
Comment (BPScott) [inline src/components/ResourceList/components/BulkActions/BulkActions.tsx]: This got called a whole load of times, so instead of computing the actionSections repeatedly, lets do it once

File: src/components/ResourceList/components/BulkActions/BulkActions.tsx [hunk 5]
```diff
@@ -317,19 +320,16 @@ class BulkActions extends React.PureComponent<CombinedProps, State> {
...
     let combinedActions: ActionListSection[] = [];
 
-    if (this.actionSections && rolledInPromotedActions.length > 0) {
-      combinedActions = [
-        {items: rolledInPromotedActions},
-        ...this.actionSections,
-      ];
-    } else if (this.actionSections) {
-      combinedActions = this.actionSections;
+    if (actionSections && rolledInPromotedActions.length > 0) {
```
Comment (BPScott) [inline src/components/ResourceList/components/BulkActions/BulkActions.tsx]: Same logic here, just prettier reformatting.

**Comment by BPScott** [comment]
Added comments to any changes that I didn't already mention as part of #2237

File: src/components/DropZone/tests/DropZone.test.tsx [hunk 6]
```diff
@@ -473,6 +473,9 @@ function setBoundingClientRect(size: keyof typeof widths) {
         left: 0,
         bottom: 0,
         right: 0,
+        x: 0,
+        y: 0,
+        toJSON() {},
```
Comment (dleroux) [inline src/components/DropZone/tests/DropZone.test.tsx]: Why is this needed?

Comment (beefchimi) [reply src/components/DropZone/tests/DropZone.test.tsx]: Probably to satisfy the type `Element['getBoundingClientRect']`:
https://github.com/Shopify/polaris-react/pull/2549/files#diff-ba67dc9a18e9f3acebe10b28d3397442R1029

Comment (BPScott) [reply src/components/DropZone/tests/DropZone.test.tsx]: TS Updated their types and added these properties https://github.com/Shopify/polaris-react/pull/2237#discussion_r344419536
